document/src/vespa/document/update/fieldupdate.h

```diff
@@ -58,23 +53,19 @@ class FieldUpdate : public vespalib::Identifiable,
      * @param update A pointer to the value update to add to this.
      * @return A pointer to this.
      */
-    FieldUpdate& addUpdate(const ValueUpdate& update) {
-        update.checkCompatibility(_field); // May throw exception.
-        _updates.push_back(ValueUpdate::CP(update.clone()));
-        return *this;
-    }
+    FieldUpdate& addUpdate(const ValueUpdate& update);
```

Comment:
To avoid having to wrap `std::move` around `FieldUpdate(foo).addUpdate(bar)` etc it's possible to use reference qualifiers to add an overload here that can only be selected by the compiler when the function is called on an rvalue (such as a stack-constructed `FieldValue` temporary). Example adding an explicit lvalue and rvalue ref qualifier to the functions (note: have not compiled this locally):

```c++
FieldUpdate& addUpdate(const ValueUpdate& update) &;
FieldUpdate&& addUpdate(const ValueUpdate& update) &&
{
    addUpdate(update);
    return std::move(*this);
}
```
Since `this` implicitly becomes an lvalue inside an rvalue function, the `addUpdate` is a forwarding function and not an infinite recursion since the rvalue function won't resolve itself as a candidate. This is also why we still have to explicitly `std::move(*this)` for the return.

A function call such as `my_rvalue_accepting_func(FieldValue(foo).addUpdate(bar).addUpdate(baz))` should then work as expected.

Reply:
Thanks, that was better.
I already had a large followup touching same code, so in order to avoid massive merge conflict, I adde that too before reverting the std::move()